test/middleware.js

```diff
@@ -47,12 +48,246 @@ describe('middleware', function() {
+      var expectedError = new Error('Bad dog!');
+      this.backend.use('readDoc', function(_request, next) {
+        return next(expectedError);
+      });
+
+      this.backend.fetch({}, 'dogs', 'fido', function(err) {
+        expect(err).to.eql(expectedError);
+        done();
+      });
+    });
+  });
+
+  describe('readDocBulk', function() {
+
+    it('is triggered when multiple documents are fetched by ids', function(done) {
+      this.backend.use('readDocBulk', function(request, next) {
+        expect(request.collection).to.eql('dogs');
+        expect(request.snapshotMap).to.have.property('fido');
+        expect(request.snapshotMap).to.have.property('spot');
+        next();
+        return done();
+      });
+
+      this.backend.fetchBulk({}, 'dogs', ['fido', 'spot'], function(err) {
+        if (err) throw(err);
+      });
+    });
+
+    it('calls back with an error that is yielded by fetchBulk', function(done) {
+      var expectedError = new Error('Bad dogs!');
+      this.backend.use('readDocBulk', function(_request, next) {
+        return next(expectedError);
+      });
+
+      this.backend.fetchBulk({}, 'dogs', ['fido', 'spot'], function(err) {
+        expect(err).to.eql(expectedError);
+        done();
+      });
+    });
+
+  });
+
+  describe('readDocs', function() {
+    beforeEach('Add fido to db', function(done) {
+      this.snapshot = {v: 1, type: 'json0', data: {age: 3}};
+      this.backend.db.commit('dogs', 'fido', {v: 0, create: {}}, this.snapshot, null, done);
+    });
+
+    ['queryFetch', 'querySubscribe'].forEach((function(queryMethod) {
+      it('is triggered when multiple documents are retrieved with ' + queryMethod, function(done) {
+        this.backend.use('readDocs', function(request, next) {
+          expect(request.collection).to.eql('dogs');
+          expect(request.snapshots).to.have.length(1);
+          expect(request.snapshots[0]).to.have.property('id', 'fido');
+          expect(request.snapshots[0]).to.have.property('data').eql({age: 3});
+          next();
+          return done();
+        });
+
+        this.backend[queryMethod]({}, 'dogs', {age: 3}, {}, function(err) {
+          if (err) throw(err);
+        });
+      });
+
+      it('calls back with an error that is yielded by ' + queryMethod, function(done) {
+        var expectedError = new Error('Bad dog!');
+        this.backend.use('readDocs', function(_request, next) {
+          return next(expectedError);
+        });
+
+        this.backend[queryMethod]({}, 'dogs', {age: 3}, {}, function(err) {
+          expect(err).to.eql(expectedError);
+          done();
+        });
+      });
+
+    }).bind(this));
+
+  });
+
   describe('submit', function() {
 
     it('gets options passed to backend.submit', function(done) {
       this.backend.use('submit', function(request, next) {
         expect(request.options).eql({testOption: true});
-        done();
+        next();
```

Comment:
Is this trying to fix a problem? It seems like an atypical pattern, but i'm trying to understand it.

Reply:
This test could have been more cleanly written as:

```
it('gets options passed to backend.submit', function(done) {
      this.backend.use('submit', function(request, next) {
        expect(request.options).eql({testOption: true});
        next();
      });
      var op = {create: {type: types.defaultType.uri}};
      var options = {testOption: true};
      this.backend.submit(null, 'dogs', 'fido', op, options, done);
    });
```

Perhaps you can use that pattern in the other tests, too